plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CacheBasedReusableConnectionFactory.java

```diff
@@ -0,0 +1,157 @@
+                        return;
+                    }
+                    try {
+                        requireNonNull(notification.getValue(), "notification.getValue() is null");
+                        notification.getValue().close();
+                    }
+                    catch (SQLException e) {
+                        throw new TrinoException(JDBC_ERROR, e);
+                    }
+                }));
+        this.connections = connections.asMap();
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    public Connection openNonReusableConnection(ConnectorSession session)
+            throws SQLException
+    {
+        return delegate.openConnection(session);
+    }
+
+    @Override
+    public Connection openConnection(ConnectorSession session)
+            throws SQLException
+    {
+        String queryId = session.getQueryId();
+        Connection connection = getConnection(session, queryId);
+        return new CachedConnection(queryId, connection);
+    }
+
+    private Connection getConnection(ConnectorSession session, String queryId)
+            throws SQLException
+    {
+        Connection connection = connections.remove(queryId);
+        if (connection != null) {
+            return connection;
+        }
+        return delegate.openConnection(session);
+    }
+
+    @Override
+    public void beginQuery(ConnectorSession session) {}
+
+    @Override
+    public void cleanupQuery(ConnectorSession session)
+    {
+        Connection connection = connections.remove(session.getQueryId());
+        if (connection != null) {
+            try {
+                connection.close();
+            }
+            catch (SQLException e) {
+                throw new TrinoException(JDBC_ERROR, e);
+            }
+        }
+    }
+
+    @Override
+    public void close()
+            throws SQLException
+    {
+        for (Connection connection : connections.values()) {
+            connection.close();
+        }
+        connections.clear();
+        delegate.close();
+    }
+
+    private final class CachedConnection
+            extends ForwardingConnection
+    {
+        private final String queryId;
+        private final Connection delegate;
+        private volatile boolean closed;
+
+        private CachedConnection(String queryId, Connection delegate)
+        {
+            this.queryId = requireNonNull(queryId, "queryId is null");
+            this.delegate = requireNonNull(delegate, "delegate is null");
+        }
+
+        @Override
+        protected Connection delegate()
+        {
+            checkState(!closed, "Connection is already closed");
+            return delegate;
+        }
+
+        @Override
+        public void close()
+                throws SQLException
+        {
+            if (closed) {
+                return;
+            }
+            closed = true;
+            connections.put(queryId, delegate);
```

Comment:
What does `session.getQueryId()` returns for `ALICE`? 

looks like this heavily relies on that metadata requests are not parallel.

and this test could be extended:
```
    @Test
    public void testConnectionIsNotShared()
            throws Exception
    {
        try (MockConnectionFactory mockConnectionFactory = new MockConnectionFactory();
                ReusableConnectionFactory connectionFactory = new CacheBasedReusableConnectionFactory(mockConnectionFactory, Duration.ofSeconds(5), 10);
                Connection ignored = connectionFactory.openConnection(ALICE)) {
            connectionFactory.openConnection(ALICE).close();
            assertThat(mockConnectionFactory.openedConnections).isEqualTo(2);
        }
    }
```